### PR TITLE
Adding support for setting multiple properties at once atomically

### PR DESCRIPTION
**Pull request by cshannon**
Here is a draft PR with my in progress changes for #2692  so it can start being reviewed.

I still need to do some more clean up (documentation and things), maybe some more tests and also the last piece of updating the shell commands to use the new features.

I plan to continue working on it later this week (probably this weekend).

I squashed the commits from #2717 and based on my PR commit on that PR.

### CODE REVIEW

**Comment by cshannon** [comment]
@ctubbsii and @dlmarion - here is my draft PR, it's still a work in progress as I state in the description but there's quite a bit here so feel free to give feedback.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:667]: How is this different than `getConfiguration` above?

**Review by dlmarion — CHANGES_REQUESTED** [review]
I did an incomplete first pass as I wanted to get clarification on the API changes before going further. I don't remember if the API methods `getTableSpecificConfiguration` and `getNamespaceSpecificConfiguration` were in the prior PR or if they are new. I see what they are doing in the implementation, but will they actually return different results?

File: server/base/src/main/java/org/apache/accumulo/server/client/ClientServiceHandler.java
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/client/ClientServiceHandler.java:345]: I would have to re-run the unit tests and switch out the call but I was getting back something like 250 properties when not filtering vs maybe 160 when filtering so there were quite a few that came back as it was all properties. Property.isValidTablePropertyKey is called when processing the updates so any property that failed that check caused a failure. The other option is to filter out properties that are not valid but figured it would be better to be explicit about it and fail vs just silently filtering.

File: core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java:237]: Same reason as the TableOperations class where it's filtering by allowed (table only) properties that can be mutated

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:667]: I just ran `TableOperationsIT.createTable` and modified it to print out the returned properties. You weren't joking. It looks like `getConfiguration(String)` was added in 2.1.0.  I wonder if that method should be renamed. We should get some other opinions here. @milleruntime @ctubbsii @EdColeman

Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:667]: @dlmarion - Yeah I figured based on that it would make a lot more sense to filter the returned properties from the server vs trying to do it later on the client side and sending all of that useless stuff hence the new method/call. I'd think either renaming the method or having a flag/argument to filter what is returned makes sense. (ie return all, return just table props, etc).

**Comment by cshannon** [comment]
Looks like #2796 was merged so I will rebase and refactor this PR based on that when I get back to this. I'm planning to continue working on this PR again this weekend. I'll continue on with updating the tests, docs, and shell code. If we determine to change how the getConfiguration() stuff works I can fix that too.

File: core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java:237]: Is this meant to return the configuration stack (default -> system -> namespace) or only the properties for the namespace?  I think we should differentiate when dealing with properties (specific prop name, values for one id) vs configuration when it is the configuration hierarchy.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:667]: If this is returning the properties from ZooKeeper that are key by id, then the naming would be getTableProperties and getNamespaceProperties - to me configuration should imply the configuration hierarchy, that hierarchy if build from properties and possibly other sources (say configuration files,...)

File: core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java:237]: This is only meant to return the configuration that can be mutated in Zookeeper for the namespace so I'm going to rename this to just getNamespaceConfiguration() as Dave suggested yesterday. The getConfiguratoin() method that's above already returns everything.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java:269]: See previous comments - is this returning configuration or properties - if just properties, then the name should reflect that.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1119]: Same as previous comment - configuration or property?

File: core/src/main/thrift/client.thrift
Comment (EdColeman) [inline core/src/main/thrift/client.thrift:335]: Same - configuration or properties?

File: core/src/main/thrift/client.thrift
Comment (EdColeman) [inline core/src/main/thrift/client.thrift:351]: Same configuration or properties

File: core/src/test/java/org/apache/accumulo/core/clientImpl/TableOperationsHelperTest.java
Comment (EdColeman) [inline core/src/test/java/org/apache/accumulo/core/clientImpl/TableOperationsHelperTest.java:165]: Same - getTableProperties() ?

File: core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java:269]: Thanks working on updates now, I will message when I am ready to have this reviewed again.

**Review by EdColeman — COMMENTED** [review]
Looks good for now.  Couple of things that I'd like to look at in more detail but generally heading in the right directions.  My  general issues / thing that I want to look at in more depth.

* naming convention, configuration vs properties
* is there duplicate functionality in handling system properties?  Some things need to be uniqque, but there may be some overlap.
* Exception handling with invalid property - currently I think we just throw IllegalStateException on first offender - not sure if this represent a change in behavior - or if it make a difference if it does.
* I did not look at the tests

**Comment by cshannon** [comment]
@EdColeman  and @dlmarion -

Ok I have updated my PR with a few changes to address the comments/discussion made as well as added/changed some tests. I rebased my PR off main and cherry-picked the commit from #2803 so that it will be easy to rebase when #2803 is merged. I then fixed up my PR to incorporate the changes made in https://github.com/EdColeman/accumulo/tree/prop_atomic_update . I also ran through the Sunny day test profile and everything passed.

**To address Ed's comments:**

- I renamed everything to be consistent so getSystemProperties(), getTableProperties(), getNamespaceProperties(), etc as we are only returning properties and not the full config.
- For system properties I updated the PR to be handled the same was as table properties and namespace properties. Same naming convention and new method on instanceOperations() etc. There may be work to do for here in another PR as stated if we want to deduplicate things
- For the exception handling I think it makes sense to fail fast on the first invalid property and throw an exception. It looks like the same IllegalStateException is thrown already on setting an invalid property as things get validated so the behavior is the same.
- For the tests I went and updated/added some more. I tried to do a search and add some tests in various areas to cover functionality (adding, removing, etc) and I also tried to add some tests for permissions. My guess is the tests may not be complete and I may need to review it and write some more edge cases tests but at least what is there should demonstrate the basic functionality.

**Other notes:**

- I reverted the changes Emily did to the Config command. It doesn't really make sense to try and mutate multiple properties at once with a command since it would blow stuff away. If we do want to have a command I would think we would need an entirely new command (not mess with the current one) and that is probably best with a follow on PR. Maybe we just have a way to script or upload a set of properties.
- The javadocs probably need some work and review for accuracy. I figured I would go back and fix up the Javadocs after the next review and when things look good before the final merge to make sure I'm not spending a lot of time on Javadocs for things that will be removed/deleted anyways.

**Comment by cshannon** [comment]
@EdColeman - I just rebased off main since I saw you merged in your PropUtil changes and pushed up the latest

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: With some changes, this code may be able to gracefully handle multiple clients concurrently modifying props.  Not completely sure, but think something like the following may work.


Create a thrift type to return version+props.

```java
class ThriftVersionedProperties {
   long version;
   Map<String, String> props;
}
```

The following code was written quickly, trying to show how we could pass a version back and forth over thrift and use that version to detect concurrent changes and retry.

```java
public void modifyProperties(String tableName, final Consumer<Map<String,String>> mapMutator)
      throws AccumuloException, AccumuloSecurityException, IllegalArgumentException,
      ConcurrentModificationException {

   while(true) {
      // make thrift call to get props+version
      ThriftVersionedProps tvp = client.getTableProperties(...);
      
      // mutate the properties
      mapMutator.accept(tvp.properties);

      try {
          // pass the modified props+version back to thrift call.  The thrift call will throw a thrift exception if the version has changed since we read props above.
          client.modifyTableProperties(...,tableName, tvp);
           break;
      } catch (PropertiesVersionChanged pvc) {
          log.debug("Properties changed since reading, retrying...");
          //TODO maybe sleep using retry object
      } catch (TableNotFoundException e) {
         throw new AccumuloException(e);
      }
   }
}
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1039]: I need to research this some more and get back to you, so the following comment is just the beginning of a discussion (feel free to ignore now).  Taking a quick look at this, i suspect checkLocalityGroups could be restructured to support setting multiple properties at once.  I'll update with a new comment after I do some more digging.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: Could this be done as a separate issue?

Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: I was going to suggest that..I think this could probably be a follow on PR as an improvement.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1039]: Same as above comment, this could probably be a follow on task as a separate issue and PR as an improvement.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: Opened https://github.com/apache/accumulo/issues/2821 with this comment as a separate issue.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: The previous ticket [included the idea of throwing ConcurrentModificationException](https://github.com/apache/accumulo/pull/2717#issuecomment-1133596040) when the server detected that the properties changed. That would require the version information suggested in #2821 be incorporated into these changes.

Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1029]: I missed that the first time around, I'll take a look at adding that version info to this PR this week and we can close out #2821 since it will be part of this instead.

**Comment by cshannon** [comment]
@ctubbsii, @EdColeman, @dlmarion  - I just pushed an updated draft PR that has changes to handle detect and handle version changes when modifying properties and is ready for review again. I added a new commit for now just to make it easier to see what my latest changes were. I will rebase/squash the commits and update the commit message when it's ready for merging.

I took @keith-turner's suggestion and went ahead and updated the `modifyProperties() ` methods to retry on concurrent modification exceptions up to 5 times before giving up and propagating the exception back to the caller of the method. We could also make this fail immediately and not retry if desired.

Let me know what you think of the latest updates.

File: test/src/main/java/org/apache/accumulo/test/util/Wait.java
Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/util/Wait.java:49]:
```suggestion
    boolean conditionSatisfied = condition.isSatisfied();
    while (!conditionSatisfied && System.currentTimeMillis() < expiry) {
      TimeUnit.MILLISECONDS.sleep(sleepMillis);
      conditionSatisfied = condition.isSatisfied();
    }
    return conditionSatisfied;
```
This util seems really nice! Just noticed this small typo while looking over it.

File: test/src/main/java/org/apache/accumulo/test/util/Wait.java
Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/util/Wait.java:25]: Could use the `TimeUnit` conversion for these instead of manually calculating seconds to millis.
```suggestion
  public static final long MAX_WAIT_MILLIS = SECONDS.toMillis(30);
```

File: test/src/main/java/org/apache/accumulo/test/util/Wait.java
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/util/Wait.java:49]: Thanks, I can fix this. The utility is actually borrowed from ActiveMQ (another project I work on heavily). Got to love the Apache licensing :)

https://github.com/apache/activemq/blob/f5f1141d36363b137c87b20f11512f9b48057c89/activemq-broker/src/test/java/org/apache/activemq/util/Wait.java

File: test/src/main/java/org/apache/accumulo/test/util/Wait.java
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/util/Wait.java]: I changed this and fixed the typo in the latest update.

File: core/src/test/java/org/apache/accumulo/core/conf/PropertyModificationUtilTest.java
Comment (dlmarion) [inline core/src/test/java/org/apache/accumulo/core/conf/PropertyModificationUtilTest.java:35]: I think the Junit5 way of doing this is to use `assertThrows`, see [example](https://github.com/apache/accumulo/blob/main/server/base/src/test/java/org/apache/accumulo/server/ServerDirsTest.java#L115)

Comment (cshannon) [reply core/src/test/java/org/apache/accumulo/core/conf/PropertyModificationUtilTest.java]: Thanks for the tip, I fixed this.

**Comment by cshannon** [comment]
@ctubbsii - This PR has been updated to remove the retry on ConcurrentModificationException

**Comment by cshannon** [comment]
I went ahead and squashed my commits into 1 (leaving Emily's) and I also rebased and removed the draft status as I think this PR is about ready to merge after a final review.

**Comment by Manno15** [comment]
> I went ahead and squashed my commits into 1 (leaving Emily's) and I also rebased and removed the draft status as I think this PR is about ready to merge after a final review.

Squashing isn't necessary usually. We squash when we merge the final PR. Squashing and force pushing can make it harder to follow the history of the ticket while reviewing.

**Comment by cshannon** [comment]
> > I went ahead and squashed my commits into 1 (leaving Emily's) and I also rebased and removed the draft status as I think this PR is about ready to merge after a final review.
> 
> 
> 
> Squashing isn't necessary usually. We squash when we merge the final PR. Squashing and force pushing can make it harder to follow the history of the ticket while reviewing. 

Ok sounds good in the future I can leave the commits and squash later on merge. I still like force pushing when just rebasing though as nothing is changing with the commit itself (usually)

**Comment by cshannon** [comment]
@Manno15 - I went ahead and restored my previous commit to make it easier to review the concurrent modification changes so the PR is back to 3 commits again.

File: test/src/main/java/org/apache/accumulo/test/util/Wait.java
Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/util/Wait.java:23]: Maybe this could be broken out into its own PR so it can be merged in separately. I think adding this class to its own PR could stand alone. I've found a spot in https://github.com/apache/accumulo/pull/2902 that could benefit from this.

**Comment by keith-turner** [comment]
> I took @keith-turner's suggestion and went ahead and updated the modifyProperties() methods to retry on concurrent modification exceptions up to 5 times before giving up and propagating the exception back to the caller of the method. 

@cshannon  where in the code does it retry 5 times? I looked around for this and did not see it.  A lot of things in Accumulo retry forever unless explicitly given a timeout.  I think retrying 5 times is hard to reason about for a user, especially someone who just wants to props to be set and are not too concerned w/ how long it takes.  They would have to write the code to retry and it would probably not be as optimal as if Accumulo was retrying (like a user could not optimally implement exponential back-off on top of code that does some sort of retrying on its own).


> I reverted the changes Emily did to the Config command. It doesn't really make sense to try and mutate multiple properties at once with a command since it would blow stuff away. If we do want to have a command I would think we would need an entirely new command (not mess with the current one) and that is probably best with a follow on PR. Maybe we just have a way to script or upload a set of properties.

I think it does make sense to update the shell command in a separate PR. It would be good to attempt that before any release that has this new API as a it will help validate the new API.  We could possibly update the shell command to support a conditional update with this new API.  Something like update props p1=v11,p2=v21, etc only if the current props are p1=v1 and p2=v2.  This would allow a user to do the following.

 * print the current props
 * create a command that updates props only if those props have the values they saw when they printed.

**Comment by EdColeman** [comment]
Regarding @keith-turner suggestion - depending on where / how that is implemented the data version of the properties can be used to tell if the  property node has changed. It may not be necessary to actually compare values.

**Comment by cshannon** [comment]
@keith-turner - The retry was removed as it was discussed and determined to not be necessary or wanted and should just fail instead. I forget when but I believe it was an offline conversation with @ctubbsii and others.

A separate PR to update the shell would be a good idea if we want to do that to keep the scope of this PR smaller as it's already a decent sized change. I could add that as a follow on once his is merged.

**Comment by keith-turner** [comment]
>Regarding @keith-turner suggestion - depending on where / how that is implemented the data version of the properties can be used to tell if the property node has changed. It may not be necessary to actually compare values. 

I suggested comparing values thinking the data version was not available through the public API.  But could still compare values through the public api.  Also the suggestion for the shell was just a WAG about how it should be done. Not exactly sure how the shell command should look.  My main point was that its really good to use a new API before releasing it.

> I think I misread your comment initially as I was thinking you meant to do the PR after this PR is merged (but before releasing 2.1) but I think you meant to go ahead and do the shell changes first.

I wasn't thinking it needed to be done before merging this PR, just before release like you were initially thinking.  One way have handled a situation like this in the past is to create a new issue that is a blocker before merging this.

> @keith-turner - The retry was removed as it was discussed and determined to not be necessary or wanted and should just fail instead. I forget when but I believe it was an offline conversation with @ctubbsii and others.

I think building in retry behavior is nice for the following reasons.

 * The concurrent modification exception (CME) is spurious and therefore hard for a user to test.  User would most likely only see it in a production system, not a test system.
 * I suspect every user who uses the API would just want it to set the props even if there was CME (its a spurious condition that has nothing to do with the data). So that means each user who uses the API will have to rewrite some sort of retry code (retry code that is hard for them to test).
 * If we build in the functionality then we can test it and write a more complex retry strategy like exponential back off that will not retry forever even when there is lot of concurrency.  We can do the hard part instead of pushing that off on every user of the API.

However I don't think implementing retry should hold up this PR. I think it would be fine to do that as a follow on if there is no opposition to it.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:687]: Do we need to say "This new method returns a Map instead of an Iterable."? If we leave this in place (in multiple places), maybe we should add something to say why?

File: server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStore.java
Comment (dlmarion) [inline server/base/src/main/java/org/apache/accumulo/server/conf/store/PropStore.java:80]:
```suggestion
   * Replaces all current properties with map provided. If a property is not included in the new map,
```

**Review by dlmarion — COMMENTED** [review]
I think this looks good. I'm not approving as I'm not super familiar with the new property code.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:687]: I don't think so, I am just going to drop that clause.

**Comment by EdColeman** [comment]
We should keep the clause about eventually consistent until we can strengthen the use of versioning information so that there is a stronger guarantee.  Its in the nature of a distributed system - but it still can be supersizing, and even causes problems in our tests if we set and then immediately expect that to be reflected across the running processes, treating it as some sort of atomic operation.

**Comment by cshannon** [comment]
> We should keep the clause about eventually consistent until we can strengthen the use of versioning information so that there is a stronger guarantee. Its in the nature of a distributed system - but it still can be supersizing, and even causes problems in our tests if we set and then immediately expect that to be reflected across the running processes, treating it as some sort of atomic operation.

That part is still there, I just removed "This new method returns a Map instead of an Iterable."

https://github.com/apache/accumulo/pull/2799/files#diff-55271a1466a3153c9fd3ed242bd3c37f38fd98c46a87ce1d2993f65d5606ca52R683-R687

**Pull request merged**
